hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/fair/FairScheduler.java

```diff
@@ -987,9 +978,19 @@ void attemptScheduling(FSSchedulerNode node) {
       }
 
       // Assign new containers...
-      // 1. Check for reserved applications
-      // 2. Schedule if there are no reservations
-
+      // 1. Ensure containers are assigned to the apps that preempted
+      // 2. Check for reserved applications
+      // 3. Schedule if there are no reservations
+
+      // Apps may wait for preempted containers
+      // We have to satisfy these first to avoid cases, when we preempt
+      // a container for A from B and C gets the preempted containers,
+      // when C does not qualify for preemption itself.
+      for (FSAppAttempt app : node.getPreemptionList()) {
```

Comment:
Also, we are not counting allocated resources/containers against the assignMultiple. We probably should so that we limit the number of non-preemption, non-reserved containers/resources appropriately.